### PR TITLE
feat: Keep `PreferencesController` in sync with keyring state

### LINKED ISSUE #3794: [preferences-controller] React to keyring state changes
The `PreferencesController` should listen for `KeyringController` state changes, and react to them appropriately. Specifically it should listen for account additions and removals, and ensure the selected address always points at an address that exists.

Today it doesn't react to these event changes because it relies upon the `KeyringController` calling the `PreferencesController` methods to update state. This is error prone and makes the `KeyringController` more complex than we'd like, and introduces a circular dependency between the controllers that would be nice to avoid.

### PR DESCRIPTION
**Pull request by Gudahtt**
## Explanation

The `PreferencesController` now listens for `KeyringController` state changes, and updates its own state in response to account removals or additions. New accounts are added to the `identities` state and given default labels. Removed accounts are removed from the `identities` state, and the `selectedAddress` is updated if it was removed.

The dependency between these two packages has been flipped; the `@metamask/preferences-controller` package now depends upon `@metamask/keyring-controller` rather than the other away around, so that the `KeyringController` state type and state change event can be accessed. The dependency the other way was just to get the types for the four `PreferencesController` methods that are passed into the `KeyringController` constructor. These types were easily inlined, and these methods will soon be removed anyway.

A `getDefaultKeyringState` export was added to the `@metamask/keyring-controller` package to make it easier to write the necessary `PreferencesController` tests.

## References

Closes #3794

#### Added
- Add `getDefaultKeyringState` function

#### Removed
- Remove `peerDependency` and `devDependency` upon `@metamask/preferences-controller`
  - This dependency was just used to access the types of four methods. Those types are now inlined instead.

#### Changed
- **BREAKING:** Keep `PreferencesController` state synchronized with `KeyringController` state
  - The `KeyringController:stateChange` event is now required by the `PreferencesController` messenger, which is a breaking change.
  - The package `@metamask/keyring-controller` has been added as a `peerDependency` and as a `devDependency`, which is a breaking change.
  - Previously the state was synchronized manually by calling `syncIdentities` or  `updateIdentities`. Calling these methods is no longer required.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/preferences-controller/src/PreferencesController.test.ts [hunk 1]
```diff
@@ -32,6 +39,189 @@ describe('PreferencesController', () => {
...
+            { accounts: ['0x00', '0x01', '0x02'], type: 'CustomKeyring' },
+          ],
+        },
+        [],
+      );
+
+      expect(controller.state.identities).toStrictEqual(identitiesState);
+    });
+
+    it('should not update selected address if it is still among identities', () => {
```
Comment (mcmire) [inline packages/preferences-controller/src/PreferencesController.test.ts:205]: This test seems to be for `updateIdentities`. Should it go in another `describe`?

**Review by mcmire — COMMENTED** [review]
One question, but otherwise looks good.

File: packages/preferences-controller/src/PreferencesController.test.ts (hunk 1 above)
Comment (Gudahtt) [reply packages/preferences-controller/src/PreferencesController.test.ts:205]: Good catch! No this belongs here, but this test was meant to trigger a state change, not call updateIdentities. These tests were copied and modified, but I forgot to modify this one I guess

Comment (Gudahtt) [reply packages/preferences-controller/src/PreferencesController.test.ts]: Fixed here: https://github.com/MetaMask/core/pull/3799/commits/bd4819d8ac2e0bf2d7de1502a7244d59a6f8fef3

**Review by mikesposito — APPROVED** [review]
LGTM!

**Comment by Gudahtt** [comment]
Merged in `main` to resolve conflicts

**Review by mikesposito — APPROVED** [review]

**Comment by Gudahtt** [comment]
Updated to use new latest keyring-controller version and bump lockfile

**Review by MajorLift — APPROVED** [review]
Looks good!

**Pull request merged**